test/core/promise/promise_fuzzer.cc

```diff
@@ -12,10 +12,306 @@
+    for (int i = 0; !done_ && activity_ != nullptr && i < msg.actions_size();
+         i++) {
+      // Do some things
+      const auto& action = msg.actions(i);
+      switch (action.action_type_case()) {
+        // Force a wakeup
+        case promise_fuzzer::Action::kForceWakeup:
+          activity_->ForceWakeup();
+          break;
+        // Cancel from the outside
+        case promise_fuzzer::Action::kCancel:
+          ExpectCancelled();
+          activity_.reset();
+          break;
+        // Flush any pending wakeups
+        case promise_fuzzer::Action::kFlushWakeup:
+          if (wakeup_ != nullptr) absl::exchange(wakeup_, nullptr)();
+          break;
+        // Drop some wakeups (external system closed?)
+        case promise_fuzzer::Action::kDropWaker: {
+          int n = action.drop_waker();
+          auto v = std::move(wakers_[n]);
+          wakers_.erase(n);
+          break;
+        }
+        // Wakeup some wakeups
+        case promise_fuzzer::Action::kAwakeWaker: {
+          int n = action.awake_waker();
+          auto v = std::move(wakers_[n]);
+          wakers_.erase(n);
+          for (auto& w : v) {
+            w.Wakeup();
+          }
+          break;
+        }
+        case promise_fuzzer::Action::ACTION_TYPE_NOT_SET:
+          break;
+      }
+    }
+    ExpectCancelled();
+    activity_.reset();
+    if (wakeup_ != nullptr) absl::exchange(wakeup_, nullptr)();
+    GPR_ASSERT(done_);
+  }
+
+ private:
+  // Schedule wakeups against the fuzzer
+  struct Scheduler {
+    Fuzzer* fuzzer;
+    // Schedule a wakeup
+    template <typename ActivityType>
+    void ScheduleWakeup(ActivityType* activity) {
+      GPR_ASSERT(activity == fuzzer->activity_.get());
+      GPR_ASSERT(fuzzer->wakeup_ == nullptr);
+      fuzzer->wakeup_ = [activity]() { activity->RunScheduledWakeup(); };
+    }
+  };
+
+  // We know that if not already finished, the status when finished will be
+  // cancelled.
+  void ExpectCancelled() {
+    if (!done_ && !expected_status_.has_value()) {
+      expected_status_ = absl::CancelledError();
+    }
+  }
+
+  // Construct a promise factory from a protobuf
+  PromiseFactory<IntHdl> MakePromiseFactory(
+      const promise_fuzzer::PromiseFactory& p) {
+    switch (p.promise_factory_type_case()) {
+      case promise_fuzzer::PromiseFactory::kPromise:
+        return [p, this](IntHdl) { return MakePromise(p.promise()); };
+      case promise_fuzzer::PromiseFactory::kLast:
+        return [](IntHdl h) { return [h]() { return h; }; };
+      case promise_fuzzer::PromiseFactory::PROMISE_FACTORY_TYPE_NOT_SET:
+        break;
+    }
+    return [](IntHdl) {
+      return []() -> Poll<IntHdl> { return std::make_shared<int>(42); };
+    };
+  }
+
+  // Construct a promise from a protobuf
+  Promise<IntHdl> MakePromise(const promise_fuzzer::Promise& p) {
+    switch (p.promise_type_case()) {
+      case promise_fuzzer::Promise::kSeq:
+        switch (p.seq().promise_factories_size()) {
```

Comment:
Could we eliminate a bunch of code here if the Seq type supported something like the builder pattern, or some other kind of decoupled construction from sequence definition?

```
std::vector<PromiseFactory> factories;
for (int i = 0; i < p.seq().promise_factories_size(); ++i) factories.push_back(MakePromiseFactory(...));
return Seq(MakePromise(p.seq().first)).all(factories);
```

Reply:
we could yeah, and we'll likely add that in the future, but it's relatively non-trivial work that i'd like to punt on for now.